### ISSUE TITLE
Find a way to add descriptions to queries and mutations

### ISSUE THREAD

**Issue by acron0** [WorksHub/leona#11]
[ch2302]

For documentation purposes, it would be nice if there was a way to add descriptions to queries and mutations so that they appear in GraphQL introspection.

**Comment by devurandom** [comment]
Would it be possible to access the docstrings of resolver functions for this?

**Comment by acron0** [comment]
That would definitely be ideal...if specs had docstrings :/ Right now I think we'll have to make do with expanding the attach fns so they can optionally accept a description.

**Comment by devurandom** [comment]
Alex said that docstrings are still on the list of things to be included in spec.alpha2. I personally would just wait for that, instead of adding a bandaid that then has to be maintained indefinitely.

**Comment by Wryhder** [comment]
> Right now I think we'll have to make do with expanding the attach fns so they can optionally accept a description.

@acron0 I would like to work on this, if no one else already is.

**Comment by acron0** [comment]
@Wryhder Yep, please do 👍

**Comment by acron0** [comment]
@kevinmungai Yes, this is basically 50% of the solution! The description then needs to be pushed into the schema as a kind of post-generation step.

**Comment by kevinmungai** [comment]
@acron0 Thanksl for the feedback. It will take me sometime before I get a working solution.

**Comment by kevinmungai** [comment]
Hi @acron0, I am having a bit of an issue. I have tried to generate a simple schema from the example on the **README** but I just can't.

## the expected
a lacinia schema should be generated using this
```clojure
(spec/def ::query-spec map?)
(spec/def ::response map?)

(-> (leona/create)
                  (leona/attach-query ::query-spec ::response (fn [context query value]))
                  (leona/compile))
```

## what am experiencing
> Execution error (IllegalArgumentException) at leona.schema/transform (schema.clj:110).
Don't know how to create ISeq from: clojure.lang.Keyword

## what I have tried
> I have tried to run the tests of `leona` but I get
Execution error (IllegalArgumentException) at leona.schema/transform (schema.clj:110).
Don't know how to create ISeq from: clojure.lang.Keyword


To be honest this is my first time trying to contribute to open source, I think I might be the issue but after 2 days of trying I have decided to ask for help. Thank you for your time.

**Comment by acron0** [comment]
@kevinmungai Don't worry :) Leona is still young, and there is still some holes in the documentation so that could be the problem here. What version of Leona are you using? And you say running the tests gives you a `IllegalArgumentException`?

**Comment by wizzytod** [comment]
Hi, @acron0 !  

May I suggest using `spec-tools` to add a description for attach-query's query-spec?
E.g.
```clojure
(s/def ::droid-query (st/spec {:spec (s/keys :req-un [::id]
                                       :opt [::appears-in])
                               :description "Query for droid"}))

(-> (leona/create)
            (leona/attach-query ::test/droid-query ::test/droid droid-resolver)
          ...)
```

**Comment by acron0** [comment]
@wizzytod Yes, I would accept PR that connects the dots to make this work

**Comment by benjamin-asdf** [comment]
1) Adding descriptions to objects via metadata on resolver would make sense to me
2) how to do descriptions for fields and args? 

```clojure
;; :leona/description metadata on a resolver or mutation
;; -> add description to the object
(let [job-resolver (fn [c q v])
      job-resolver-with-meta
      (with-meta job-resolver {:leona/description "foo"})
      compiled-schema (-> (leona/create)
                          (leona/attach-query ::job-input :wh/job job-resolver-with-meta)
                          (leona/compile))]
  (-> compiled-schema
      :generated :objects :Job :description))
"foo"


;; Fields and args can also have descriptions, for completeness we would want a way to provide those?
;; not sure if metada conceptually makes sense, but we could add a :descriptions key to compile opts?

(leona/compile
 m
 {:descriptions {:wh/job "a job"
                 :wh.company/name {:fields "company name"
                                   :args "company name arg"}
                 :wh.job/id {:fields "the id of a job"
                             :args "the id of a job"}}})


```

Update, thoughts: 

Instead of args to `compile` it probably makes more sense that it is another `attach-` function. 
Also for 1) we probably want that on the level of generating, not compiling.
